Chunk naming with hashing

guybedford opened the pull request:
As discussed in #1995, this PR updates the code splitting API to the following:

```js
rollup.rollup({
  input: {
    'entryOne.js': 'path/to/entry1.js',
    'entryTwo.js': 'path/to/entry2.js'
  }, // while still supporting the array form
  output: {
    dir: 'dist',
    chunkNames: 'chunk-[hash].js'
  }
});
```

This allows custom names for both chunks and inputs.

I've maintained preserveModules in the process - and some relative-pathing bugs in the implementation of preserveModules should be fixed here as well.

The only replacement variable we have in chunk names is currently `[hash]` where the default chunk naming scheme is exactly `chunk-[hash]` now. Future variable replacements could be defined here as well.

I've included hash.js from npm for this - the total build size increases from around 710kb unminified to 727kb unminified with the build here.

The source hash is calculated only when needed in code splitting - so single file builds aren't affected. Additionally builds that don't use hashes in chunk naming won't get the cost of hash calculation either. 

To handle circular references etc the hash of a source is the hash of its rendered output before applying finalizers combined with the hash of the addons and the hash of all dependency sources (including the names of externals). Hopefully this should be quite resilient, while remaining relatively constant when source changes aren't present.

Other refactorings in the process include:
* Fixing up the watcher to use a single cache API, and exposing this new cache object in the main top-level Rollup API. This will allow us to extend the cache with new properties in future.
* Removing the object and array utility functions (happy to revert this too if it seems like overkill)
* Minor adjustments to sourceMappingURL
* Creating a new test file that tracks the configuration options lists so this can be updated in one place.

In terms of the output API, I went with the following:

```js
interface OuptputChunk {
  code: string;
  map: SourceMap;
  imports: string[];
  exports: string[];
  modules: string[];
};

// single-file case:
rollup.rollup(...): {
  // --- these can possibly be deprecated or reconsidered: ---
  imports: string[];
  exports: string[];
  modules: ModuleJSON[];

  // new cache object
  cache: { modules }

  generate: () => OutputChunk;
  write: () => OutputChunk;
}

// code-splitting case:
rollup.rollup(...): {
  // new cache object
  cache: { modules }

  generate: () => { [name: string]: OutputChunk };
  write: () => { [name: string]: OutputChunk };
}
```

We can still consider whether generate and write should match exactly, and if we really do want to consider the comment re deprecations there.

Note the OutputChunk doesn't have the id of the chunk as a property, just as the key, although we have this with code splitting currently. This could possibly be added back too.

guybedford commented:
Another thing that could be worth considering is if we should by default name entry point files by their hashes as well - eg `main1-sdf0s9d8f.js` instead of `main1.js`. This would provide far-future-expire suitable builds, but does requires users to work out which entry point becomes which name for creating their initial imports.

guybedford commented:
Also a canonical example test can be found at https://github.com/rollup/rollup/pull/2068/files#diff-9ae0b14a24dc086e196cc1e1507af161.

guybedford commented:
The only annoying thing with this is that the `transformBundle` hook isn't integrated into the hashing function because the hash is computed before finalization.

The reason for this is because of circular references 🤷‍♂️

guybedford commented:
I'll aim to see if there's something we can do about this hashing problem with `transformBundle` as we think about `transformChunk` implementations.

lukastaegert commented:
> The table of contents is too big for display.

Github is not really helping with this review...

lukastaegert commented:
Some initial findings: On first try, I did not get this to work via CLI + config file. I think some adjustments in the `bin` folder will be necessary to support the new API. For instance in `bin/src/run/build.ts`, it attempts to join my input object.

I also get some type errors in my IDE there that for some reason do not seem to prevent a successful build. Weird.

lukastaegert commented:
> Another thing that could be worth considering is if we should by default name entry point files by their hashes as well - eg main1-sdf0s9d8f.js instead of main1.js

Could you explain this a little further? In which situations do you want to add hashes? Is this about the `input` as array case or about other situations?

Not sure I like the idea. Would there be a way to get rid of the hashes? Or maybe for this situation we need an `entryPointNames` option to provide a template for entry points as well. There is still an open question if there is a way to create more than one set of chunks with different naming conventions (e.g. extensions `.js` vs. `.mjs` or `.es.js` etc.).

guybedford commented:
> Some initial findings: On first try, I did not get this to work via CLI + config file. I think some adjustments in the bin folder will be necessary to support the new API. For instance in bin/src/run/build.ts, it attempts to join my input object.

I haven't done a CLI interface for this yet... let me look into this.

> Could you explain this a little further? In which situations do you want to add hashes? Is this about the input as array case or about other situations?

Yes exactly this is for the default array case. The benefit for users is knowing that the output folder can be pushed up with far-future expires as the names are all uniquely hashed. But I guess this isn't necessary as a default - and users can easily configure it for themselves if they want.

Note that input aliases support hash variables too - `input: { 'entryA-[hash].mjs': 'path/to/entry.js' }`.

lukastaegert commented:
On second thought in the case where `input` is an array
* entry points are renamed or moved around anyway, i.e. put into a flat folder even though they are at different nesting levels and then deconflicted
* there is no way to control this behaviour (maybe there should be a way?)
* this option is still experimental
* there is no safe way to add hashes afterwards as entry points may be imported by other entry points

so maybe adding hashes is a reasonable idea. Though I expect most people to switch to the new API anyway. Maybe we should even consider deprecating the array form and removing it once code-splitting is no longer experimental?

lukastaegert commented on src/Chunk.ts:
```diff
@@ -673,160 +688,219 @@ export default class Chunk {
+		if (ext !== '.js' && ext !== '.mjs') {
+			pattern += ext;
+			ext = '.js';
+		}
```
A reasonable approach but I would expect people to know what they are doing if they specify an extension. Extensions are not part of the JavaScript spec anyway. So I would prefer the following behaviour:
* No extension provided: Either add `.js` or - add no extension at all! Open for discussion here. Otherwise there is actually no way to create a chunk without an extension (with `/bin/rollup` we even create an extension-less file ourselves)
* An extension is provided: Do not add another extension as it quite possible this is intentional and there may be other extensions beyond (m)js that could make sense for JavaScript.

guybedford commented:
Thanks for the feedback @lukastaegert.

I've fixed up the CLI handling around these cases now, and supported an aliasing form in the CLI of `rollup entryAlias=./path/to/entryA.js --experimentalCodeSplitting --dir out`.

I've also taken your suggestion and implemented an `entryNames` option that works exactly like `chunkNames` while removing automatic extension adding (see the latest commit at https://github.com/rollup/rollup/pull/2068/commits/fbcee789229c3967d57e1558db32ab879144d25c).

The default is now `chunkNames: '[alias]-[hash].js'` and `entryNames: '[alias].js'`, where the extension is full customizable. The `[alias]` replacement is taken as `chunk` by default for chunks (when we introduce manual chunks this can be set to the manual chunk name), and for entry points the alias is the key in the map for the object form and the basename without the extension for the array form.

Hopefully that addresses the feedback so far! Also if you wanted to get the other smaller PRs in before this I'm happy to rebase this to other changes.

lukastaegert commented:
> Also if you wanted to get the other smaller PRs in before this I'm happy to rebase this to other changes.

If you would prefer the smaller PRs to be reviewed first, I can do that! Will not have much time before tomorrow for more reviewing, though.

guybedford commented:
@lukastaegert sure, if limited for time please rather focus review time on this one then.

guybedford commented:
//cc @jthoms1 ☝

jthoms1 commented:
I think that generating the the hash names within all entries and chunks would be preferred.  Thanks!

lukastaegert commented:
I agree (now), hashing everything by default definitely makes sense.

lukastaegert commented on bin/src/run/index.ts:
```diff
@@ -19,8 +20,23 @@ export default function runRollup(command: any) {
 
-	if (command.output && command.output.dir) {
-		command.input = command._;
+	if (command.dir) {
```
One thing I noticed not directly related to this but which we might fix here as well: Apart from directly providing multiple file names on command line, people might also just want to use the `--input` option multiple times. This is actually supported by our command line parser which will turn the input option into an array in this case but rollup will not handle this correctly. It should be easy to transform this case properly here as well.

guybedford replied:
It seems odd that we support both `rollup x.js` and `rollup --input x.js` as meaning the same thing... perhaps we should even consider deprecating the `--input` option then, as `rollup x.js y.js z.js --dir out` seems preferable to `rollup --input x.js --input y.js --input z.js --dir out`?

lukastaegert replied:
Yes and no. When there are many options and several inputs, it can help readability if you can precede each input with `--input`.

guybedford replied:
Sure, I've added support for this.

guybedford commented on src/Chunk.ts:
```diff
@@ -673,160 +688,244 @@ export default class Chunk {
+						let alias = basename(this.entryModule.id);
+						if (alias.endsWith('.js')) alias = alias.substr(0, alias.length - 3);
+						else if (alias.endsWith('.mjs')) alias = alias.substr(0, alias.length - 4);
+						return alias;
```
Some developers like to use dots in paths to indicate hierarchy - `page.component.thing`. Assuming extensions for `.` will affect these kinds of scenarios. Also the `alias` shouldn't be considered itself a file path.

In terms of validating the `alias` further it could be worth thinking about ensuring it doesn't support `..`, `.` or `\` characters etc. with some validation.

guybedford commented on src/Chunk.ts:
```diff
@@ -673,160 +688,244 @@ export default class Chunk {
+				let ext = extname(outName);
+				if (ext === '.js' || ext === '.mjs')
+					outName = outName.substr(0, outName.length - ext.length);
+				else ext = '';
```
Again, I'm -1 to supporting other extensions. There are only two supported extensions for the JS language.

If a user really needed this, I'd rather allow registering custom js extensions.

lukastaegert replied:
I am ok with leaving it this way but then I would prefer if you put the extensions into a constant for the three different places using them to make it easier to find them later.

Then (in another PR), we could add an option e.g. `extensions:` to register a set of extensions to replace this constant that
* will be considered extensions of JavaScript files when generating names
* can be omitted when importing files (and finally make everyone happy)

The default could then be `['.js' , '.mjs']`. Though this would diverge from our previous stance on this. But since we have a preference for certain extensions now anyway, maybe it is time.

guybedford replied:
Included.

guybedford commented:
> Great updates! Really like the new alias concept as well the way how you can specify an alias via the command line!

This came from your suggestion at https://github.com/rollup/rollup/pull/2068#issuecomment-374748044 :)

> maybe we should not add any hashes by default

I may actually tend to agree as well in that it gets tricky to write tutorials when the path is going to be unpredictable and change between builds.

@jthoms1 I won't mind if we backtrack on this and ask that you just use `entryPointNames: '[alias]-[hash].js'` for this use case in your configuration.

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -379,10 +366,12 @@ export default function rollup(
-				const result: OutputChunk = {
-					imports: chunk.getImportIds(),
-					exports: chunk.getExportNames(),
-					modules: chunk.getJsonModules(),
```
Just for curiosity: What is the reason you switched `modules` to plain ids instead of complex objects with meta information? Performance?
I am definitely ok with this. Should there actually be demand for this additional information in the future, one idea for the interface might be to create additional meta-information on-demand by providing a getter here instead of a pre-made object. This could even be done in the single-chunk case in a somewhat downward compatible way by using actual getters.

guybedford replied:
The main reason is simply reducing "encouraged" API surface area. This is also why I created a new `cache` property to contain the full modules object on the Rollup output - so that users start to think of that ModuleJSON as something more internal to the cache than directly user-facing.

lukastaegert commented on bin/src/run/index.ts:
```diff
@@ -19,8 +21,22 @@ export default function runRollup(command: any) {
-		command.input = command._;
+	if (command.dir) {
+		if (!command._.some((input: string) => input.indexOf('=') !== -1)) {
+			command.input = command._;
```
Maybe we should use a similar check as in the second case so that when `--input` is used, it is not overwritten with an empty array? BTW just noticed another advantage of using ' --input`: When adding input files after boolean flags, not using `--input` will confuse the command parser.

guybedford replied:
Sure.

lukastaegert commented on src/rollup/index.ts:
```diff
@@ -488,36 +467,78 @@ export default function rollup(
+												if (chunk.isEntryModuleFacade) {
+													pattern = outputOptions.entryNames || '[alias].js';
+												} else {
+													pattern = outputOptions.chunkNames || '[alias]-[hash].js';
```
As discussed, maybe we should remove the hash from the default here.

guybedford replied:
This hash case is for chunks that won't be loaded by users only - `chunk-sd8fs9f.js`. Entry points are still `entry.js`, effectively only using the hashes on the "internal chunk names".

Perhaps in the manual chunks PR we could discuss an option here to remove the hashes by default - either by separating `manualChunkNames` or by making the alias itself include the hash, although that kind of defeats the nice mental model :P

I thought we could keep it for these cases over using the numbering (`chunk.js`, `chunk2.js`, ...), as if anything hashes probably make it easier to keep track of chunk and chunk changes for users anyway.

guybedford commented on src/ast/nodes/ExportDefaultDeclaration.ts:
```diff
@@ -68,7 +67,7 @@ export default class ExportDefaultDeclaration extends NodeBase {
 		);
 	}
 
-	render(code: MagicString, options: RenderOptions, { start, end }: NodeRenderOptions = BLANK) {
```
Ok I'll add back the `BLANK` constant.

guybedford commented on src/utils/sourceMappingURL.ts:
```diff
@@ -1,8 +1,5 @@
 // this looks ridiculous, but it prevents sourcemap tooling from mistaking
 // this for an actual sourceMappingURL
-let SOURCEMAPPING_URL = 'sourceMa';
```
Sure, will revert this commit.

guybedford commented:
(added a third to avoid whitespace issues in tests)

guybedford commented:
I've added a duplicate entry point error as well now.

guybedford commented:
Just noticed a dynamic import test case bug here... please don't merge until I've pushed a fix.

lukastaegert commented:
Great improvements, looks good 👍
